lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -2134,8 +2134,17 @@ void DifferentiableActivityInfo::propagateUseful(
   if (auto *ai = dyn_cast<ApplyInst>(inst)) {
     if (isWithoutDerivative(ai->getCallee()))
       return;
-    for (auto arg : ai->getArgumentsWithoutIndirectResults())
-      setUsefulAndPropagateToOperands(arg, i);
+    // Propagate usefulness to non-`inout` arguments.
+    // Skip `inout` arguments to avoid propagating usefulness from results to
+    // non-useful `inout` arguments (representing results).
```

Comment:
Note: skipping `inout` arguments feels somewhat ad-hoc. Semantically, `inout` arguments *are* formal arguments, so it makes sense to propagate usefulness to from `apply` results to `inout` argument operands.

Reply:
Is skipping inout arguments correct? There are certainly situations where inouts are only used as results, and it's correct in these cases, but there are also situations where inouts are also used as inputs.